File: Packages/RideSharingFeature/Sources/RideSharingFeature/DesignSystem/LiiveMapMarker.swift
import SwiftUI

public struct LiiveMapMarker: View {
    public enum Kind { case car, origin, destination, transfer }

    let kind: Kind
    let label: String?

    public init(kind: Kind, label: String? = nil) {
        self.kind = kind
        self.label = label
    }

    private var color: Color {
        switch kind {
        case .car: return LiiveColor.accent
        case .origin: return LiiveColor.success
        case .destination: return LiiveColor.danger
        case .transfer: return LiiveColor.warning
        }
    }

    private var icon: String {
        switch kind {
        case .car: return "car.fill"
        case .origin: return "location.fill"
        case .destination: return "mappin.circle.fill"
        case .transfer: return "arrow.triangle.swap"
        }
    }

    public var body: some View {
        VStack(spacing: LiiveMapMarkerLayout.markerGap) {
            if kind == .origin {
                originDot
            } else {
                pinMarker
            }
            if let label, !label.isEmpty {
                labelTag(label)
            }
        }
        .fixedSize()
    }

    private var originDot: some View {
        Circle()
            .fill(color)
            .frame(width: LiiveMapMarkerLayout.dotSize, height: LiiveMapMarkerLayout.dotSize)
            .overlay(
                Circle().stroke(
                    LiiveMapMarkerLayout.outlineColor,
                    lineWidth: LiiveMapMarkerLayout.dotStrokeWidth
                )
            )
            .liiveShadow(.pin)
    }

    private var pinMarker: some View {
        ZStack(alignment: .bottom) {
            PointerTail(color: color)
                .frame(width: LiiveMapMarkerLayout.pinTailSize, height: LiiveMapMarkerLayout.pinTailSize)
                .offset(y: LiiveMapMarkerLayout.pinTailOffset)
            Circle()
                .fill(color)
                .frame(width: LiiveMapMarkerLayout.pinSize, height: LiiveMapMarkerLayout.pinSize)
                .overlay(
                    Circle().stroke(
                        LiiveMapMarkerLayout.outlineColor,
                        lineWidth: LiiveMapMarkerLayout.pinStrokeWidth
                    )
                )
                .liiveShadow(.pin)
            Image(systemName: icon)
                .font(.system(size: LiiveMapMarkerLayout.glyphSize, weight: .bold))
                .foregroundColor(LiiveMapMarkerLayout.outlineColor)
        }
        .frame(width: LiiveMapMarkerLayout.pinSize, height: LiiveMapMarkerLayout.pinSize)
    }

    private func labelTag(_ label: String) -> some View {
        Text(label)
            .font(LiiveFont.caption1.weight(.semibold))
            .foregroundColor(LiiveColor.text)
            .lineLimit(1)
            .padding(.horizontal, LiiveMapMarkerLayout.labelHorizontalPadding)
            .padding(.vertical, LiiveMapMarkerLayout.labelVerticalPadding)
            .background(LiiveColor.surface)
            .clipShape(Capsule())
            .overlay(alignment: .bottom) {
                Capsule()
                    .fill(color)
                    .frame(height: LiiveMapMarkerLayout.labelIndicatorHeight)
            }
            .liiveShadow(.small)
    }
}

private enum LiiveMapMarkerLayout {
    static let markerGap = LiiveSpacing.xs
    static let dotSize = LiiveSpacing.l + LiiveSpacing.xs2
    static let dotStrokeWidth = LiiveSpacing.xs - LiiveSpacing.xs2 / 2
    static let pinSize = LiiveControl.md - LiiveSpacing.xs - LiiveSpacing.xs2
    static let glyphSize = dotSize
    static let pinTailSize = LiiveSpacing.m
    static let pinTailOffset = LiiveSpacing.xs + LiiveSpacing.xs2 / 2
    static let pinStrokeWidth = LiiveSpacing.xs2 + LiiveSpacing.xs2 / 4
    static let labelHorizontalPadding = LiiveSpacing.s
    static let labelVerticalPadding = LiiveSpacing.xs2
    static let labelIndicatorHeight = LiiveSpacing.xs2
    static let pointerRotationDegrees = 45.0
    static let outlineColor = Color.white
}

private struct PointerTail: View {
    let color: Color

    var body: some View {
        Rectangle()
            .fill(color)
            .overlay(alignment: .trailing) {
                Rectangle()
                    .fill(LiiveMapMarkerLayout.outlineColor)
                    .frame(width: LiiveMapMarkerLayout.pinStrokeWidth)
            }
            .overlay(alignment: .bottom) {
                Rectangle()
                    .fill(LiiveMapMarkerLayout.outlineColor)
                    .frame(height: LiiveMapMarkerLayout.pinStrokeWidth)
            }
            .rotationEffect(.degrees(LiiveMapMarkerLayout.pointerRotationDegrees))
    }
}

File: android/app/src/main/java/com/liive/ride/designsystem/LiiveMapMarker.kt
package com.liive.ride.designsystem

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.res.painterResource

@Composable
fun LiiveMapMarker(kind: MapMarkerKind, label: String? = null) {
    val c = LiiveTheme.colors
    val color = when (kind) {
        MapMarkerKind.Car -> c.accent
        MapMarkerKind.Origin -> c.success
        MapMarkerKind.Destination -> c.danger
        MapMarkerKind.Transfer -> c.warning
    }
    val icon = when (kind) {
        MapMarkerKind.Car -> RideIcons.Car
        MapMarkerKind.Origin, MapMarkerKind.Destination -> RideIcons.LocationOn
        MapMarkerKind.Transfer -> RideIcons.SwapHoriz
    }

    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(LiiveMapMarkerLayout.MarkerGap),
        modifier = Modifier.widthIn(max = LiiveMapMarkerLayout.MaxLabelWidth)
    ) {
        if (kind == MapMarkerKind.Origin) {
            Box(
                Modifier
                    .size(LiiveMapMarkerLayout.DotSize)
                    .shadow(LiiveElevation.pin, CircleShape)
                    .clip(CircleShape)
                    .background(color)
                    .border(LiiveMapMarkerLayout.DotStrokeWidth, LiiveMapMarkerLayout.OutlineColor, CircleShape)
            )
        } else {
            Box(
                Modifier.size(LiiveMapMarkerLayout.PinSize),
                contentAlignment = Alignment.Center
            ) {
                PointerTail(
                    color = color,
                    modifier = Modifier
                        .align(Alignment.BottomCenter)
                        .offset(y = LiiveMapMarkerLayout.PinTailOffset)
                )
                Box(
                    Modifier
                        .size(LiiveMapMarkerLayout.PinSize)
                        .shadow(LiiveElevation.pin, CircleShape)
                        .clip(CircleShape)
                        .background(color)
                        .border(
                            LiiveMapMarkerLayout.PinStrokeWidth,
                            LiiveMapMarkerLayout.OutlineColor,
                            CircleShape
                        ),
                    contentAlignment = Alignment.Center
                ) {
                    Icon(
                        painterResource(icon),
                        null,
                        tint = LiiveMapMarkerLayout.OutlineColor,
                        modifier = Modifier.size(LiiveMapMarkerLayout.GlyphSize)
                    )
                }
            }
        }
        if (!label.isNullOrEmpty()) {
            MarkerLabel(label = label, color = color)
        }
    }
}

@Composable
private fun PointerTail(color: Color, modifier: Modifier = Modifier) {
    Canvas(modifier.size(LiiveMapMarkerLayout.TailCanvasSize)) {
        val edge = LiiveMapMarkerLayout.TailEdgeSize.toPx()
        val strokeWidth = LiiveMapMarkerLayout.PinStrokeWidth.toPx()
        val left = (size.width - edge) / 2f
        val top = (size.height - edge) / 2f
        val right = left + edge
        val bottom = top + edge

        rotate(degrees = LiiveMapMarkerLayout.PointerRotationDegrees, pivot = center) {
            drawRect(color, topLeft = Offset(left, top), size = Size(edge, edge))
            drawLine(
                LiiveMapMarkerLayout.OutlineColor,
                Offset(right, top),
                Offset(right, bottom),
                strokeWidth = strokeWidth
            )
            drawLine(
                LiiveMapMarkerLayout.OutlineColor,
                Offset(left, bottom),
                Offset(right, bottom),
                strokeWidth = strokeWidth
            )
        }
    }
}

@Composable
private fun MarkerLabel(label: String, color: Color) {
    val c = LiiveTheme.colors

    Box(
        Modifier
            .shadow(LiiveElevation.small, LiiveRadius.full)
            .clip(LiiveRadius.full)
            .background(c.surface)
    ) {
        Text(
            label,
            color = c.text,
            style = MaterialTheme.typography.labelMedium,
            maxLines = 1,
            modifier = Modifier.padding(
                horizontal = LiiveMapMarkerLayout.LabelHorizontalPadding,
                vertical = LiiveMapMarkerLayout.LabelVerticalPadding
            )
        )
        Box(
            Modifier
                .align(Alignment.BottomCenter)
                .fillMaxWidth()
                .height(LiiveMapMarkerLayout.LabelIndicatorHeight)
                .background(color)
        )
    }
}

private object LiiveMapMarkerLayout {
    val MarkerGap = LiiveSpacing.xs
    val MaxLabelWidth = LiiveSpacing.huge + LiiveSpacing.huge + LiiveSpacing.huge +
        LiiveSpacing.xxxl - LiiveSpacing.xs2
    val DotSize = LiiveSpacing.l + LiiveSpacing.xs2
    val DotStrokeWidth = LiiveSpacing.xs - LiiveSpacing.xs2 / 2
    val PinSize = LiiveControl.md - LiiveSpacing.xs - LiiveSpacing.xs2
    val GlyphSize = DotSize
    val TailCanvasSize = DotSize
    val TailEdgeSize = LiiveSpacing.m
    val PinTailOffset = LiiveSpacing.s - LiiveSpacing.xs2 / 4
    val PinStrokeWidth = LiiveSpacing.xs2 + LiiveSpacing.xs2 / 4
    val LabelHorizontalPadding = LiiveSpacing.s
    val LabelVerticalPadding = LiiveSpacing.xs2
    val LabelIndicatorHeight = LiiveSpacing.xs2
    const val PointerRotationDegrees = 45f
    val OutlineColor = Color.White
}

enum class MapMarkerKind { Car, Origin, Destination, Transfer }
